File: GameFramePro/Assets/Scripts/Core/Base/Module/ResourcesManager/ResourcsLoad/ResourcesLoadAssetRecord.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System;

namespace GameFramePro.ResourcesEx
{
    /// <summary>
    /// 加载Resources 时候的记录
    /// </summary>
    #if UNITY_EDITOR
    [System.Serializable]
#endif
    public class ResourcesLoadAssetRecord : ILoadAssetRecord
    {

#if UNITY_EDITOR
        #region Show
        public string Debug_AssetUrl = string.Empty;
        public int Debug_InstanceID  = 0;
        public int Debug_ReferenceCount  = 0;
        public LoadedAssetTypeEnum Debug_AssetLoadedType  = LoadedAssetTypeEnum.None;
        public UnityEngine.Object Debug_TargetAsset  = null;
        public float Debug_RemainTimeToBeDelete  = 0;
        public long Debug_MarkToDeleteTime  = 0;
        public IAssetManager Debug_BelongAssetManager  = null;


        public void UpdateData()
        {
            Debug_InstanceID = InstanceID;
            Debug_AssetUrl = AssetUrl;
            Debug_ReferenceCount = ReferenceCount;
            Debug_AssetLoadedType = AssetLoadedType;
            Debug_TargetAsset = TargetAsset;
            Debug_RemainTimeToBeDelete = RemainTimeToBeDelete;
            Debug_MarkToDeleteTime = MarkToDeleteTime;
            Debug_BelongAssetManager = BelongAssetManager;
        }

        #endregion
#endif

        public int InstanceID { get; protected set; } = 0;
        public string AssetUrl { get; protected set; } = string.Empty;
        public int ReferenceCount { get; protected set; } = 0;
        public LoadedAssetTypeEnum AssetLoadedType { get; protected set; } = LoadedAssetTypeEnum.None;
        public UnityEngine.Object TargetAsset { get; protected set; } = null;
        public float RemainTimeToBeDelete { get; protected set; } = 0;
        public long MarkToDeleteTime { get; protected set; } = 0;
        public IAssetManager BelongAssetManager { get; protected set; } = null;


        #region 构造函数& 设置
        public ResourcesLoadAssetRecord()
        {

        }
        public ResourcesLoadAssetRecord(string assetPath,  LoadedAssetTypeEnum typeEnum, UnityEngine.Object asset, IAssetManager manager)
        {
            Initial(assetPath, typeEnum, asset, manager);
        }

        public void Initial(string assetPath,  LoadedAssetTypeEnum typeEnum, UnityEngine.Object asset, IAssetManager manager)
        {
            Debug.Assert(manager != null);
            Debug.Assert(asset != null);

            AssetUrl = assetPath;
            InstanceID = asset.GetInstanceID();
            ReferenceCount = 0;
            AssetLoadedType = typeEnum;
            BelongAssetManager = manager;
            TargetAsset = asset;
            RemainTimeToBeDelete = BelongAssetManager.MaxAliveTimeAfterNoReference;
        }

        #endregion



        public void AddReference()
        {
            if (TargetAsset == null)
            {
                Debug.LogErrorFormat("资源{0}  已经被卸载了, 无法增加引用", AssetUrl);
                ReferenceCount = 0;
            }
            else
            {
                ++ReferenceCount;
            }
            NotifyReferenceChange();
        }

        public void ReduceReference(bool isforceDelete = false)
        {
            if (TargetAsset == null)
            {
                Debug.LogErrorFormat("资源{0}  已经被卸载了, 无法减少引用", AssetUrl);
                ReferenceCount = 0;
            }
            else
            {
                if (isforceDelete)
                    ReferenceCount = 0;
                else
                    --ReferenceCount;
                if (ReferenceCount == 0)
                    MarkToDeleteTime = DateTime.UtcNow.ToTimestamp_Millisecond();
            }
            NotifyReferenceChange();
        }

        public bool TimeTick(float tickTime)
        {
            if (TargetAsset == null)
            {
                ReferenceCount = 0;
                NotifyReferenceChange();
                return false;
            }

            RemainTimeToBeDelete = RemainTimeToBeDelete - tickTime;
            return RemainTimeToBeDelete > 0f;
        }


        public void NotifyNoReference()
        {
            if (TargetAsset == null) return;
            GameObject go = TargetAsset as GameObject;
            if (go != null)
            {
                go.SetActive(false);
            }
        }

        public bool NotifyReReference()
        {
            ReferenceCount = 1;
            return TargetAsset != null;
        }

        public void NotifyReleaseRecord()
        {
            BelongAssetManager.NotifyAssetRelease(this);

            AssetUrl = null;
            ReferenceCount = 0;
            AssetLoadedType = LoadedAssetTypeEnum.None;
            BelongAssetManager = null;
            TargetAsset = null;
        }

        public void NotifyReferenceChange()
        {
            BelongAssetManager.NotifyAssetReferenceChange(this);
        }


    }
}

File: GameFramePro/Assets/.lpproj/Apis/Unity/CS/GameFramePro/ResourcesEx/ResourcesLoadAssetRecord.lua
---@class CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord : CS.System.Object
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord = {}

---@field public CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_AssetUrl : CS.System.String
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_AssetUrl = nil

---@field public CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_InstanceID : CS.System.Int32
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_InstanceID = nil

---@field public CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_ReferenceCount : CS.System.Int32
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_ReferenceCount = nil

---@field public CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_AssetLoadedType : CS.GameFramePro.ResourcesEx.LoadedAssetTypeEnum
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_AssetLoadedType = nil

---@field public CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_TargetAsset : CS.UnityEngine.Object
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_TargetAsset = nil

---@field public CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_RemainTimeToBeDelete : CS.System.Single
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_RemainTimeToBeDelete = nil

---@field public CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_MarkToDeleteTime : CS.System.Int64
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_MarkToDeleteTime = nil

---@field public CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_BelongAssetManager : CS.GameFramePro.ResourcesEx.IAssetManager
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.Debug_BelongAssetManager = nil

---@property readwrite CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.InstanceID : CS.System.Int32
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.InstanceID = nil

---@property readwrite CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.AssetUrl : CS.System.String
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.AssetUrl = nil

---@property readwrite CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.ReferenceCount : CS.System.Int32
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.ReferenceCount = nil

---@property readwrite CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.AssetLoadedType : CS.GameFramePro.ResourcesEx.LoadedAssetTypeEnum
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.AssetLoadedType = nil

---@property readwrite CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.TargetAsset : CS.UnityEngine.Object
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.TargetAsset = nil

---@property readwrite CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.RemainTimeToBeDelete : CS.System.Single
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.RemainTimeToBeDelete = nil

---@property readwrite CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.MarkToDeleteTime : CS.System.Int64
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.MarkToDeleteTime = nil

---@property readwrite CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.BelongAssetManager : CS.GameFramePro.ResourcesEx.IAssetManager
CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord.BelongAssetManager = nil

---@return CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord
function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord()
end

---@param assetPath : CS.System.String
---@param typeEnum : CS.GameFramePro.ResourcesEx.LoadedAssetTypeEnum
---@param asset : CS.UnityEngine.Object
---@param manager : CS.GameFramePro.ResourcesEx.IAssetManager
---@return CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord
function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord(assetPath, typeEnum, asset, manager)
end

function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:UpdateData()
end

---@param assetPath : CS.System.String
---@param typeEnum : CS.GameFramePro.ResourcesEx.LoadedAssetTypeEnum
---@param asset : CS.UnityEngine.Object
---@param manager : CS.GameFramePro.ResourcesEx.IAssetManager
function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:Initial(assetPath, typeEnum, asset, manager)
end

function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:AddReference()
end

---@param isforceDelete : CS.System.Boolean
function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:ReduceReference(isforceDelete)
end

---@param tickTime : CS.System.Single
---@return CS.System.Boolean
function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:TimeTick(tickTime)
end

function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:NotifyNoReference()
end

---@return CS.System.Boolean
function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:NotifyReReference()
end

function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:NotifyReleaseRecord()
end

function CS.GameFramePro.ResourcesEx.ResourcesLoadAssetRecord:NotifyReferenceChange()
end